modules/core/src/main/java/org/apache/ignite/internal/QueryMXBeanImpl.java

```diff
@@ -0,0 +1,150 @@
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.util.typedef.T2;
+import org.apache.ignite.internal.util.typedef.internal.A;
+import org.apache.ignite.lang.IgniteClosure;
+import org.apache.ignite.mxbean.QueryMXBean;
+import org.apache.ignite.resources.IgniteInstanceResource;
+
+import static org.apache.ignite.internal.sql.command.SqlKillQueryCommand.parseGlobalQueryId;
+
+/**
+ * QueryMXBean implementation.
+ */
+public class QueryMXBeanImpl implements QueryMXBean {
+    /** Global query id format. */
+    public static final String EXPECTED_GLOBAL_QRY_ID_FORMAT = "Global query id should have format " +
+        "'{node_id}_{query_id}', e.g. '6fa749ee-7cf8-4635-be10-36a1c75267a7_54321'";
+
+    /** Kernal context. */
+    private final GridKernalContext ctx;
+
+    /** Logger. */
+    private final IgniteLogger log;
+
+    /**
+     * @param ctx Context.
+     */
+    public QueryMXBeanImpl(GridKernalContext ctx) {
+        this.ctx = ctx;
+        this.log = ctx.log(QueryMXBeanImpl.class);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void cancelContinuous(String originNodeId, String routineId) {
+        A.notNullOrEmpty(originNodeId, "originNodeId");
+        A.notNullOrEmpty(routineId, "routineId");
+
+        if (log.isInfoEnabled())
+            log.info("Killing continuous query[routineId=" + routineId + ",originNodeId=" + originNodeId + ']');
+
+        cancelContinuous(UUID.fromString(originNodeId), UUID.fromString(routineId));
+    }
+
+    /** {@inheritDoc} */
+    @Override public void cancelSQL(String id) {
+        A.notNull(id, "id");
+
+        if (log.isInfoEnabled())
+            log.info("Killing sql query[id=" + id + ']');
+
+        try {
+            T2<UUID, Long> ids = parseGlobalQueryId(id);
+
+            if (ids == null)
+                throw new IllegalArgumentException("Expected global query id. " + EXPECTED_GLOBAL_QRY_ID_FORMAT);
+
+            cancelSQL(ids.get1(), ids.get2());
+        }
+        catch (IgniteException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Executes scan query cancel on all cluster nodes.
+     *
+     * @param originNodeId Originating node id.
+     */
+    public void cancelSQL(UUID originNodeId, long qryId) {
+        ctx.grid().compute(ctx.grid().cluster().forNodeId(originNodeId))
+            .broadcast(new CancelSQLOnInitiator(), qryId);
+    }
+
+    /**
+     * Kills continuous query by the identifier.
+     *
+     * @param originNodeId Originating node id.
+     * @param routineId Routine id.
+     */
+    public void cancelContinuous(UUID originNodeId, UUID routineId) {
+        ctx.grid().compute(ctx.grid().cluster().forNodeId(originNodeId))
+            .broadcast(new CancelContinuousOnInitiator(), routineId);
+    }
+
+    /**
+     * Cancel SQL on initiator closure.
+     */
+    private static class CancelSQLOnInitiator implements IgniteClosure<Long, Void> {
```

Comment:
i bit confused about naming ... this class knows nothing about initiator, it`s just a broadcast task, probably we need to rename it ? but it up to you

Reply:
I prefer to keep it as is to avoid discrepancies with AI sources